### PR TITLE
Adding documentation of top-level import opened and constructors

### PR DESCRIPTION
**Pull request by davidcok**

### CODE REVIEW

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 8b39e3c510f1e64d8beb3f9f36cb087fc90cbbbf
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=b730e416-1831-4ba6-8db6-74db846e699a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: bcb20207cc7eb6fae0d21b20f4d9d6185396e634
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=15aafb7a-d1b1-48b5-bda3-060c24fbe1a3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: b12eb70fd08f74bdf08c7f6c98abf0a21485314e
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=0307d35a-2a30-408d-920f-01a31c2c2fbf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 3c449fb1e154bfa255ac6738f6b13c54cee64f87
* Result: SUCCEEDED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=0ea33ca3-a837-4130-8581-57e24bd77dfc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: Docs/DafnyRef/DafnyRef.mdk [hunk 1]
```diff
@@ -1005,21 +1008,33 @@ module Mod {
...
-importing module. There is no wya to refer to the parent module, only
+importing module. There is no way to refer to the parent module, only
 sibling modules (and their submodules).
 
+Import statements may occur at the top-level as well. There they serve
+simply as a way to give a new name, perhaps
+a shorthand name, to a module. For example,
+
+```
+include "MyModule.dfy" // declares module MyModule
```
Comment (RustanLeino) [inline Docs/DafnyRef/DafnyRef.mdk]: Replace this `include` with the definition of module `MyModule` itself. The `include` feature is just a way to, essentially, add more files to the Dafny command line, and as such is unrelated to imports.

File: Docs/DafnyRef/DafnyRef.mdk [hunk 2]
```diff
@@ -1066,6 +1081,19 @@ is not an error, however, as long as you don't attempt to use members
 with common names. The `opened` keyword can be used with any kind of
 `import` declaration, including the module abstraction form.
 
+An `import opened` can occur at the top-level as well. For example,
+```
+include "MyModule.dfy" // declares module MyModule
```
Comment (RustanLeino) [inline Docs/DafnyRef/DafnyRef.mdk]: Same comment as above: replace this `include` with a literal module definition.

File: Docs/DafnyRef/DafnyRef.mdk [hunk 3]
```diff
@@ -3727,34 +3755,83 @@ Note that when there is no body, Dafny assumes that the *ensures*
...
+To write structured object-oriented programs, one often relies on 
+objects being constructed only in certain ways.  For this purpose, Dafny
 provides _constructor (method)s_, which are a restricted form of
 initialization methods.  A constructor is declared with the keyword
-`constructor` instead of `method`.  When a class contains a
-constructor, every call to `new` for that class must be accompanied
-with a call to one of the constructors.  Moreover, a constructor
+`constructor` instead of `method`.  
+Every call to `new` for a class must be accompanied
+by a call to one of its constructors.  Moreover, a constructor
```
Comment (RustanLeino) [inline Docs/DafnyRef/DafnyRef.mdk]: The previous text is correct, the new text is not. A call to `new` must be accompanied by a call to one of the class's constructors only if the class declares any constructors at all.

File: Docs/DafnyRef/DafnyRef.mdk [hunk 4]
```diff
@@ -3727,34 +3755,83 @@ Note that when there is no body, Dafny assumes that the *ensures*
...
 ```
-When invoking this constructor, the "`.`" is dropped, as in:
+The named constructor is invoked as
+```
+  Item i := new Item.I(42);
+```
+The anonymous constructor is invoked as
 ```
-m := new Item(45, 29);
+  Item i := new Item(42,43);
```
Comment (RustanLeino) [inline Docs/DafnyRef/DafnyRef.mdk]: Add space between arguments `42,` and `43`.

File: Docs/DafnyRef/DafnyRef.mdk [hunk 5]
```diff
@@ -3727,34 +3755,83 @@ Note that when there is no body, Dafny assumes that the *ensures*
...
+If there is no `new;` statement, the entire body is the initialization phase. 
+This initialization phase is intended to initialize field variables. 
+In this phase, uses of the object reference `this` are restricted;
+a program may use `this`
+
+ - as the receiver on the LHS, 
+ - as the entire RHS of an assignment to a field of `this`, 
+ - and as a member of a set on the RHS that is being assigned to a field of `this`.
+
+Furthermore, `const` fields may only be assigned to (once) in an initialization phase
```
Comment (RustanLeino) [inline Docs/DafnyRef/DafnyRef.mdk]: You can actually assign to a `const` multiple times in the initialization phase.

File: Docs/DafnyRef/DafnyRef.mdk [hunk 6]
```diff
@@ -3727,34 +3755,83 @@ Note that when there is no body, Dafny assumes that the *ensures*
...
+of their parent class, and then only if they do not already have an initialization
```
Comment (RustanLeino) [inline Docs/DafnyRef/DafnyRef.mdk]: I suggest "enclosing class" instead of "parent class", since the latter may suggest something about subclassing.

**Comment by robin-aws** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: Dafny
* Commit ID: 4fc79d693cdaa462742c96beed91b145c53210e0
* Result: FAILED
* [Build Logs](https://g85ugprat7.execute-api.us-west-2.amazonaws.com/Prod/buildlogs?key=b9fc055c-b0b3-421b-a000-518290bb6154%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Pull request closed without merging**